pxtlib/webusb.ts

```diff
@@ -421,104 +447,101 @@ namespace pxt.usb {
-                        this.epOut = this.altIface.endpoints.filter(e => e.direction == "out")[0]
-                        Util.assert(this.epIn.packetSize == 64);
-                        Util.assert(this.epOut.packetSize == 64);
-                    } else {
-                        this.log("using ctrl pipe")
                     }
-                    this.log("claim interface")
-                    return dev.claimInterface(iface.interfaceNumber)
-                })
-                .then(() => {
-                    this.log("device ready")
-                    this.lastKnownDeviceSerialNumber = this.dev.serialNumber;
-                    this.ready = true;
-                    if (isHF2)
-                        this.readLoop();
-                    if (this.onConnectionChanged)
-                        this.onConnectionChanged();
-                })
+                }
+                return false;
+            }
+            this.log("got " + dev.configurations[0].interfaces.length + " interfaces");
+            const matching = dev.configurations[0].interfaces.filter(matchesFilters);
+            let iface = matching[matching.length - 1];
+            this.log(`${matching.length} matching interfaces; picking ${iface ? "#" + iface.interfaceNumber : "n/a"}`);
+            if (!iface)
+                this.error("cannot find supported USB interface");
+            this.altIface = iface.alternates[0];
+            this.iface = iface;
+            if (this.altIface.endpoints.length) {
+                this.log("using dedicated endpoints");
+                this.epIn = this.altIface.endpoints.filter(e => e.direction == "in")[0];
+                this.epOut = this.altIface.endpoints.filter(e => e.direction == "out")[0];
+                Util.assert(this.epIn.packetSize == 64);
+                Util.assert(this.epOut.packetSize == 64);
+            } else {
+                this.log("using ctrl pipe");
+            }
+            this.log("claim interface");
+            await dev.claimInterface(iface.interfaceNumber);
+            this.log("device ready");
+            this.lastKnownDeviceSerialNumber = this.dev.serialNumber;
+            this.ready = true;
+            if (isHF2) {
+                // just starting, not waiting on it.
+                this.readLoop();
+            }
+            this.onConnectionChanged?.();
         }
     }
 
-    export function pairAsync(): Promise<boolean> {
-        return ((navigator as any).usb.requestDevice({
-            filters: filters
-        }) as Promise<USBDevice>)
-            .then(dev => !!dev)
-            .catch(e => {
-                // user cancelled
-                if (e.name == "NotFoundError")
-                    return undefined;
-                throw e;
-            })
+    export async function pairAsync(): Promise<boolean> {
+        try {
+            const dev = await navigator.usb.requestDevice({
```

Comment:
Could safe continuation be needed here? There are a few other instances in the file as well.
```suggestion
            const dev = await navigator?.usb?.requestDevice({
```

Reply:
https://github.com/microsoft/pxt/blob/dev/jwunderl/async-ify-webusb-cleanup/pxtlib/webusb.ts#L608 is where it'll be checked overall, but adding in here for clarity sake ~